ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala

```diff
@@ -243,6 +265,88 @@ object KeyValueConsumption {
     )
   }
 
+  private[kvutils] def outOfTimeBoundsEntryToUpdate(
+      recordTime: Timestamp,
+      outOfTimeBoundsEntry: DamlOutOfTimeBoundsEntry): Option[Update] = {
+    val wrappedLogEntry = outOfTimeBoundsEntry.getEntry
+    val duplicateUntilMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasDuplicateUntil,
+      outOfTimeBoundsEntry.getDuplicateUntil)
+    val tooEarlyUntilMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasTooEarlyUntil,
+      outOfTimeBoundsEntry.getTooEarlyUntil)
+    val tooLateFromMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasTooLateFrom,
+      outOfTimeBoundsEntry.getTooLateFrom)
+    val deduplicated = duplicateUntilMaybe.exists(recordTime <= _)
+    val tooEarly = tooEarlyUntilMaybe.exists(recordTime < _)
+    val tooLate = tooLateFromMaybe.exists(recordTime > _)
+    val invalidRecordTime = tooEarly || tooLate
+
+    wrappedLogEntry.getPayloadCase match {
+      case _ if deduplicated =>
+        // We don't emit updates for deduplicated submissions.
+        None
+
+      case DamlLogEntry.PayloadCase.TRANSACTION_REJECTION_ENTRY if invalidRecordTime =>
+        val transactionRejectionEntry = wrappedLogEntry.getTransactionRejectionEntry
+        val reason = (tooEarlyUntilMaybe, tooLateFromMaybe) match {
+          case (Some(lowerBound), Some(upperBound)) =>
+            s"Record time $recordTime outside of range [$lowerBound, $upperBound]"
+          case _ =>
+            "Ledger time outside of valid range"
```

Comment:
If the single-bound cases are valid ones (as it seems from the tests), I think we should produce better messages here; if not, this is a bug and the tests should be changed accordingly.

Reply:
Added specific reasons for single-bound cases.